File: backup-typescript/src/utils/totp.ts
import speakeasy from 'speakeasy';
import QRCode from 'qrcode';
import { ITotpToken } from '../types';

/**
 * TOTP工具类
 * 基于提供的GitHub库 https://github.com/jaden/totp-generator 的概念
 * 使用speakeasy库实现TOTP功能
 */
export class TotpManager {
  
  /**
   * 生成TOTP令牌
   * @param secret TOTP密钥
   * @returns TOTP令牌和剩余时间
   */
  static generateToken(secret: string): ITotpToken {
    // 使用speakeasy生成TOTP令牌
    const token = speakeasy.totp({
      secret: secret,
      encoding: 'base32',
      time: Math.floor(Date.now() / 1000),
      step: 30, // 30秒步长
      digits: 6  // 6位数字
    });

    // 计算剩余时间
    const timeRemaining = 30 - (Math.floor(Date.now() / 1000) % 30);

    return {
      token,
      timeRemaining
    };
  }

  /**
   * 验证TOTP令牌
   * @param token 用户输入的令牌
   * @param secret TOTP密钥
   * @param window 时间窗口(允许前后几个步长)
   * @returns 验证结果
   */
  static verifyToken(token: string, secret: string, window: number = 1): boolean {
    return speakeasy.totp.verify({
      secret: secret,
      encoding: 'base32',
      token: token,
      step: 30,
      digits: 6,
      window: window
    });
  }

  /**
   * 生成新的TOTP密钥
   * @param name 账号名称
   * @param issuer 发行者名称
   * @returns 密钥信息
   */
  static generateSecret(name: string, issuer: string = 'GitHub Manager'): {
    secret: string;
    base32: string;
    otpauth_url: string;
  } {
    const secret = speakeasy.generateSecret({
      name: name,
      issuer: issuer,
      length: 32
    });

    return {
      secret: secret.ascii,
      base32: secret.base32,
      otpauth_url: secret.otpauth_url || ''
    };
  }

  /**
   * 生成QR码
   * @param otpauth_url OTP认证URL
   * @returns Base64编码的QR码图片
   */
  static async generateQRCode(otpauth_url: string): Promise<string> {
    try {
      const qrCodeData = await QRCode.toDataURL(otpauth_url);
      return qrCodeData;
    } catch (error) {
      throw new Error('Failed to generate QR code: ' + error);
    }
  }

  /**
   * 根据用户提供的密钥生成TOTP（兼容GitHub的TOTP格式）
   * @param secret 用户提供的密钥字符串
   * @returns TOTP令牌信息
   */
  static generateFromUserSecret(secret: string): ITotpToken {
    // 清理密钥字符串（移除空格和特殊字符）
    const cleanSecret = secret.replace(/\s+/g, '').toUpperCase();
    
    try {
      return this.generateToken(cleanSecret);
    } catch (error) {
      throw new Error('Invalid TOTP secret format');
    }
  }

  /**
   * 格式化显示TOTP令牌
   * @param token TOTP令牌
   * @returns 格式化的令牌字符串 (例如: 123 456)
   */
  static formatToken(token: string): string {
    if (token.length === 6) {
      return `${token.substring(0, 3)} ${token.substring(3)}`;
    }
    return token;
  }

  /**
   * 获取下一个令牌的时间
   * @returns 下一个令牌生成的时间戳
   */
  static getNextTokenTime(): number {
    const now = Math.floor(Date.now() / 1000);
    const nextStep = Math.ceil(now / 30) * 30;
    return nextStep * 1000;
  }

  /**
   * 批量生成多个账号的TOTP
   * @param accounts 账号列表，包含密钥
   * @returns 所有账号的TOTP令牌
   */
  static generateMultipleTokens(accounts: Array<{id: number, username: string, totpSecret: string}>): Array<{
    id: number;
    username: string;
    token: ITotpToken;
  }> {
    return accounts.map(account => {
      try {
        const token = this.generateFromUserSecret(account.totpSecret);
        return {
          id: account.id,
          username: account.username,
          token
        };
      } catch (error) {
        return {
          id: account.id,
          username: account.username,
          token: {
            token: 'ERROR',
            timeRemaining: 0
          }
        };
      }
    });
  }
}

File: backend/utils/totp.py
"""
TOTP工具模块
"""

import pyotp
import time
from typing import Dict


def generate_totp_token(secret: str) -> Dict[str, str]:
    """
    生成TOTP令牌
    
    Args:
        secret: TOTP密钥
        
    Returns:
        包含令牌信息的字典
    """
    try:
        # 创建TOTP对象
        totp = pyotp.TOTP(secret)
        
        # 生成当前令牌
        token = totp.now()
        
        # 计算剩余时间
        current_time = int(time.time())
        time_remaining = 30 - (current_time % 30)
        
        # 格式化令牌（添加空格）
        formatted_token = f"{token[:3]} {token[3:]}"
        
        return {
            "token": token,
            "time_remaining": time_remaining,
            "formatted_token": formatted_token
        }
    except Exception as e:
        raise ValueError(f"生成TOTP令牌失败: {str(e)}")


def verify_totp_token(secret: str, token: str) -> bool:
    """
    验证TOTP令牌
    
    Args:
        secret: TOTP密钥
        token: 要验证的令牌
        
    Returns:
        验证结果
    """
    try:
        totp = pyotp.TOTP(secret)
        return totp.verify(token, valid_window=1)  # 允许前后30秒的窗口
    except Exception:
        return False


def validate_totp_secret(secret: str) -> bool:
    """
    验证TOTP密钥格式是否正确
    
    Args:
        secret: TOTP密钥
        
    Returns:
        验证结果
    """
    try:
        # 尝试创建TOTP对象来验证密钥格式
        pyotp.TOTP(secret)
        return True
    except Exception:
        return False